model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java

```diff
@@ -129,12 +130,23 @@ public boolean isMetadataInitialized() {
     }
 
     /**
-     * In case of any update on entity, we want to update the entityVerion
+     * In case of any update on entity, we want to update the entityVersion
      * to current one.
+     * This includes downgrading from a future version of Keycloak, as the entityVersion must match the JSON
+     * and the additional tables this version writes.
+     *
+     * The @PreUpdate makes JPA calling this when any attribute of this entity has been changed and is about to be persisted
+     * .
+     * Must be called explicitly when a child (like the attributes) is modified, as that doesn't trigger @PreUpdate.
+     * Still, the entityVersion needs to match the application's SUPPORTED_VERSION_CLIENT as the structure and contents
+     * of the additional tables might depend on the version.
+     *
+     * Future versions of this method might restrict downgrading to downgrade only from the next version.
      */
-    private void checkEntityVersionForUpdate() {
+    @PreUpdate
+    private void updateEntityVersion() {
         Integer ev = getEntityVersion();
-        if (ev != null && ev < SUPPORTED_VERSION_CLIENT) {
+        if (ev != null && !Objects.equals(ev, SUPPORTED_VERSION_CLIENT)) {
```

Comment:
Thank you @ahus1. I like the way how it's done.

I think we might restrict downgrading the entity_version only from next keycloak version already. wdyt?
 
+1 for moving the method and javadoc to "future" `JpaRootEntiy`. And also from my POV it's ok to call the method when updating the attributes.

Reply:
It sounds reasonable, +1, thank you!

Reply:
+1 to @ahus1 's proposal